superbench/benchmarks/micro_benchmarks/directx_gemm_flops_performance/GPUCore.cpp

```diff
@@ -0,0 +1,507 @@
+    {
+        ComPtr<ID3D12Debug> debugController;
+        if (SUCCEEDED(D3D12GetDebugInterface(IID_PPV_ARGS(&debugController)))) {
+            debugController->EnableDebugLayer();
+
+            // Enable additional debug layers.
+            dxgiFactoryFlags |= DXGI_CREATE_FACTORY_DEBUG;
+        }
+    }
+#endif
+
+    ComPtr<IDXGIFactory4> factory;
+    ThrowIfFailed(CreateDXGIFactory2(dxgiFactoryFlags, IID_PPV_ARGS(&factory)));
+
+    ComPtr<IDXGIAdapter1> hardwareAdapter;
+    GetHardwareAdapter(factory.Get(), &hardwareAdapter);
+
+    // Create GPU device object.
+    ThrowIfFailed(D3D12CreateDevice(hardwareAdapter.Get(), D3D_FEATURE_LEVEL_11_0, IID_PPV_ARGS(&m_device)));
+
+    DML_CREATE_DEVICE_FLAGS dmlCreateDeviceFlags = DML_CREATE_DEVICE_FLAG_NONE;
+
+#if defined(_DEBUG)
+    // If the project is in a debug build, then enable the Direct3D 12 debug layer.
+    // This is optional (starting in DML_FEATURE_LEVEL_5_2) but strongly recommended!
+
+    // If the project is in a debug build, then enable debugging via DirectML debug layers with this flag.
+    dmlCreateDeviceFlags |= DML_CREATE_DEVICE_FLAG_DEBUG;
+#endif
+
+    ThrowIfFailed(DMLCreateDevice(m_device.Get(), dmlCreateDeviceFlags, IID_PPV_ARGS(m_dmlDevice.GetAddressOf())));
+
+    D3D12_COMMAND_QUEUE_DESC queueDesc;
+    // Initialize command queue.
+    ZeroMemory(&queueDesc, sizeof(queueDesc));
+
+    // Describe and create the command queue.
+    queueDesc.Flags = D3D12_COMMAND_QUEUE_FLAG_NONE;
+    queueDesc.Type = D3D12_COMMAND_LIST_TYPE_DIRECT;
+
+    D3D12_COMMAND_QUEUE_DESC cqd3 = {};
+    cqd3.Type = D3D12_COMMAND_LIST_TYPE_DIRECT;
+    m_device->CreateCommandQueue(&cqd3, IID_PPV_ARGS(&m_commandQueue));
+
+    m_device->CreateCommandAllocator(D3D12_COMMAND_LIST_TYPE_DIRECT, IID_PPV_ARGS(&m_commandAllocator));
+
+    m_device->CreateCommandList(0, D3D12_COMMAND_LIST_TYPE_DIRECT, m_commandAllocator.Get(), nullptr,
+                                IID_PPV_ARGS(&m_commandList));
+
+    ThrowIfFailed(m_commandList->Close());
+    // Reset the command list to prep for initialization commands.
+    ThrowIfFailed(m_commandList->Reset(m_commandAllocator.Get(), nullptr));
+
+    // Create fence.
+    ThrowIfFailed(m_device->CreateFence(0, D3D12_FENCE_FLAG_NONE, IID_PPV_ARGS(&m_fence)));
+    m_currentFence = 1;
+    // Create an event handle to use for GPU synchronization.
+    m_eventHandle = CreateEvent(0, false, false, 0);
+}
+
+inline UINT64 DMLCalcBufferTensorSize(DML_TENSOR_DATA_TYPE dataType, UINT tensorElementCount) {
+    UINT elementSizeInBytes = 0;
+    switch (dataType) {
+    case DML_TENSOR_DATA_TYPE_FLOAT32:
+
+        elementSizeInBytes = 4;
+        break;
+
+    case DML_TENSOR_DATA_TYPE_FLOAT16:
+
+        elementSizeInBytes = 2;
+        break;
+
+    default:
+        return 0; // Invalid data type
+    }
+
+    UINT64 minimumImpliedSizeInBytes = 0;
+
+    // Round up to the nearest 4 bytes.
```

Comment:
It's not nearest 4 byte based on this code.

Reply:
refer to https://github.com/microsoft/DirectML/blob/master/Libraries/DirectMLX.h

Reply:
But the code is incorrect for this comment.

Reply:
the comment is copied from the library code